model-engine/model_engine_server/inference/vllm/vllm_batch.py

```diff
@@ -0,0 +1,341 @@
+
+    async for i, res in results_generator:
+        if isinstance(res, AsyncGenerator):
+            continue
+        outputs[i] = res
+        bar.update(1)
+    return outputs
+
+
+async def generate_completions(
+    engine: AsyncEngineClient, request: BatchCompletionContent
+) -> Union[List[Optional[CompletionV1Output]], List[Optional[CompletionResponse]]]:
+    if isinstance(request, CreateBatchCompletionsV1RequestContent):
+        return await generate_v1_completions(engine, request)
+    elif isinstance(request, List):
+        return await generate_v2_completions(engine, request)
+    else:
+        assert_never(request)
+
+
+async def init_engine(
+    model: str,
+    request: CreateBatchCompletionsEngineRequest,
+) -> AsyncEngineClient:
+    global openai_serving_chat
+    global openai_serving_completion
+
+    if request.attention_backend is not None:
+        os.environ["ATTENTION_BACKEND"] = request.attention_backend
+
+    engine_args = AsyncEngineArgs(
+        model=model,
+        tensor_parallel_size=request.model_cfg.num_shards,
+        seed=request.model_cfg.seed or 0,
+        disable_log_requests=True,
+        gpu_memory_utilization=request.max_gpu_memory_utilization or 0.9,
+        max_model_len=request.model_cfg.max_context_length,
+    )
+
+    async_engine_client = AsyncLLMEngine.from_engine_args(engine_args)
+    model_config = await async_engine_client.get_model_config()
+    served_model_names = [model]
+
+    openai_serving_chat = OpenAIServingChat(
+        async_engine_client,
+        model_config,
+        served_model_names,
+        response_role=request.model_cfg.response_role or "assistant",
+        lora_modules=None,
+        prompt_adapters=None,
+        request_logger=None,
+        chat_template=None,
+    )
+
+    openai_serving_completion = OpenAIServingCompletion(
+        async_engine_client,
+        model_config,
+        served_model_names,
+        lora_modules=None,
+        prompt_adapters=None,
+        request_logger=None,
+    )
+
+    return async_engine_client
+
+
+def overwrite_request(request: Dict[str, Any], model: str) -> Dict[str, Any]:
+    request["model"] = model
+    request["stream"] = False
+    return request
+
+
+def load_batch_content(
+    request: CreateBatchCompletionsEngineRequest,
+) -> BatchCompletionContent:
+    content = request.content
+    if content is None:
+        with smart_open.open(request.input_data_path, "r") as f:
+            data = json.load(f)
+            try:
+                return CreateBatchCompletionsV1RequestContent.model_validate(data)
+            except ValidationError:
+                model = get_model_name(request.model_cfg)
+                if not isinstance(data, List) or len(data) == 0:
+                    raise ValueError("Input data must be a list of completion requests")
+
+                sample = data[0]
+                # Kinda jank, but we first serialize to FilteredCompletionV2Request before serializing to
```

Comment:
i think it's okay but may consider adding a type so we readily know what the type is?

Reply:
The type is FilteredCompletionV2Request that we cast to vLLMs CompletionRequest. I think vLLMs pydantic spec is too strict, but that's out of our control